### ISSUE TITLE
Ripple API support

### ISSUE THREAD

**Issue by MarkusTeufelberger** [timmolter/XChange#181]
While this might be a bit more difficult to manage (Ripple allows several issuers per currency, so there might be the need to set up multiple "exchanges" e.g. "USD.Bitstamp <--> BTC.Bitstamp" and "USD.Bitstamp <--> BTC.Weexchange") it might be good to have the prices on Ripple for both display and active trading available in XChange.

Documentation of API commands is available at https://ripple.com/wiki/RPC_API and the interface to ripple would be either a local server ("rippled") or any public one (a safe one to use would be s1.ripple.com). Commands can be sent via POST or (preferrably) websockets.

I don't do much Java but at least I could help with any questions that arise. :)

**Comment by ww3456** [comment]
I've now added order entry, order cancel and balance query [here](https://github.com/ww3456/XChange/commit/efe8bc5c0f8d19659fee55c4a273229ee7274b08) . However, I've only briefly tested it so don't feel it is quite ready for a pull request yet. However, the order cancel is causing a couple of rescu warnings in the log file:
```
WARN  [si.mazi.rescu.RestMethodMetadata] DELETE request declared as consuming method body as application/json. While body is allowed, it should be ignored by the server. Is this intended?
WARN  [si.mazi.rescu.RestInvocation] DELETE request will contain a body. While this is allowed, the body should be ignored by the server. Is this intended?
```
Since the Ripple REST API requires a json body in the order cancel DELETE request (https://github.com/ripple/ripple-rest#cancel-order), can @mmazi or anyone suggest the best way of suppressing, or avoiding these warnings?

**Comment by mmazi** [comment]
The reason for the warnings is that ripple-rest doesn't conform to the HTTP protocol. Now that you know this, you can safely ignore the warnings :) The warnings are intended for the (rescu) users to start bugging the server (ripple-rest in this case) developers to make a HTTP-protocol-conforming server.

An alternative to ignoring the warnings is to configure logging differently (eg. set threshold for `si.mazi.rescu.RestInvocation` to ERROR)  -- note that I **don't recommend doing this** since it will turn off other logging that may be important.

Also note that the RestMethodMetadata warning will be logged once (per client instance), while the RestInvocation warning will be logged with every order cancel request.

Bottom line: I think the current behavior with the logged warnings makes sense, considering that people should not write servers that don't conform to protocols, and considering that some clients will simply fail (or, I believe, issue a POST instead of a DELETE, etc.) in such cases.

**Comment by ww3456** [comment]
Thanks for the explanation, if it starts creating too much noise I'll look into adding an application logback filter for the specific message text. 

@ripple would it be possible to make your ripple-rest API HTTP protocol compliant for order cancel requests? i.e. not use a DELETE message requiring a message body.

**Comment by MarkusTeufelberger** [comment]
It seems to conform: http://tools.ietf.org/rfcmarkup?doc=7231#section-4.3.5

"A payload within a DELETE request message has no defined semantics; sending a payload body on a DELETE request might cause some existing implementations to reject the request."

**Comment by mmazi** [comment]
From [the HTTP 1.1 spec](http://www.w3.org/Protocols/rfc2616/rfc2616.html):

[4.3](http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.3): 
>  [I]f the request method does not include defined semantics for an entity-body, then the message-body SHOULD be ignored when handling the request. 

Section [9.7 DELETE](http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.7) does not mention "entity-body" or anything similar.  I interpret this as there not being any defined semantics for the body for DELETE, hence (by 4.3) the server should ignore the message body. The ripple-rest server does not (cannot) ignore the message body (it *requires* the data therein) so I believe it to be non-conformant.

Note that *the client* that sends a DELETE with body *does* conform to the spec: sending the body is allowed, but the server should respond in the same way no matter what the body is.

**Comment by MarkusTeufelberger** [comment]
From the same document you linked: "This document has been superseded. In 2014, RFC2616 was replaced by multiple RFCs (7230-7237)."

I linked you the current version of RFC 7231.

I also think it's stupid to put a body in a DELETE request (and the secret key should be transferred as a header instead), also there is a clear warning in 7231 that some existing implementations may break or just ignore anything there. Feel free to file bug reports at https://github.com/ripple/ripple-rest/issues.

**Comment by timmolter** [comment]
done.
